content/2015.05.cliquet-permissions-en.rst

```diff
@@ -0,0 +1,535 @@
+to implement permissions management in a generic data storage system**
+
+The problem at hand
+===================
+
+The problem we're facing is simple: data is stored online and we need a way to
+share this data with other people.
+
+Looking at the use cases, it's possible to split our users in two categories:
+
+- "Final" users (you);
+- Applications which interact on their behalf.
+
+Different parties don't have the same rights: some should be able to read,
+others should write, yet another should be able to create new records, and all
+this permission handling should be fine grained: it should be possible to read
+one record but not the other, for instance.
+
+We started from the fact that no available solution were offering an appealing
+solution to these needs.
+
+
+A vocabulary problem
+====================
+
+The principal problem that got in our way during our thinking was the
+vocabulary.
+
+
+Principals
+----------
+
+A *principal*, is an entity that can be authenticated by a computer system.
+It is the actor who does the action. [#]_
+
+It could either be someone, a computer, a service or a group of any of such
+things. This is conceptually wider than the classical "user id*".
+
+Permissions are then associated to these *principals*.
+
+For instance, an user is identified in an unique way when he connects to the
+authentication system, which role is to define a list of *principals* for the
+authenticating user.
+
+.. [#] To read more about *principals* :
+       https://en.wikipedia.org/wiki/Principal_%28computer_security%29
+
+
+The difference between a role and a group
+-----------------------------------------
+
+From scratch, it's not easy to precisely define the difference between these two
+concepts which allow to associate permissions to a group of *principals*. [#]_
+
+The difference is mainly semantic, but we can see a difference in the "direction"
+of the relationship between these two concepts.
+
+- A role is a list of permissions associated to a *principal*;
+- A group is a list of *principals* we can associate to a permission.
+
+.. [#] More information at:
+       http://stackoverflow.com/questions/7770728/group-vs-role-any-real-difference
+
+
+The difference between permission, ACL and ACE
+----------------------------------------------
+
+.. epigraph::
+
+  An ACL is a list of Access Control Entry (ACE) enabling or disabling
+  acces rights to a person or a group.
+
+  -- https://en.wikipedia.org/wiki/Access_Control_List
+
+I would even say, in our case, "to a *principal*". For instance:
+
+.. code-block:: text
+
+    create_record: alexis,remy,tarek
+
+This ACE gives the ``create`` permission on the ``record`` object to the users
+Tarek, Rémy and Alexis.
+
+
+Permissions delegation
+======================
+
+Imagine the following example, where an user stores two types of data online:
+
+- contacts ;
+- a list of tasks, which can be associated to her contacts.
```

Comment:
Also, that's probably Shitstorm Compliant Ubernitpicking™, but I'm more comfortable with using *they* instead of *her*; *they* include everybody by design.